### PR TITLE
BDOG-2678 Automatically update year changes in headers

### PR DESCRIPTION
**Pull request by RikSherman**
Only add a new header not update the year on existing headers.

### CODE REVIEW

File: src/sbt-test/sbt-auto-build/test-source-generation-older-sbt/build.sbt [hunk 1]
```diff
@@ -15,7 +15,10 @@ lazy val root = (project in file("."))
         !updated.contains(s"Copyright ${java.time.Year.now()} HM Revenue & Customs")
       }
       val failed = sources.collect { case source if noHeader(source) => source }
-      if (failed.nonEmpty)
-        sys.error(s"Header was not added to ${failed.mkString(",")}")
+      val oldDateUpdated = Source.fromFile("app/OldController.scala").getLines().mkString.contains(s"Copyright ${java.time.Year.now()} HM Revenue & Customs")
+      if (failed.nonEmpty || oldDateUpdated) {
```
Comment (colin-lamed) [inline src/sbt-test/sbt-auto-build/test-source-generation-older-sbt/build.sbt:19]: I think handling the two error case separately will be clearer:
```scala
if (failed.nonEmpty)
  sys.error(s"Header was not added to ${failed.mkString(",")}")
if (oldDateUpdated)
 sys.error(""Header date was updated in app/OldController.scala")
```

File: src/sbt-test/sbt-auto-build/test-source-generation-older-sbt/build.sbt [hunk 2]
```diff
@@ -15,7 +15,10 @@ lazy val root = (project in file("."))
...
+        val message = s"${if (failed.nonEmpty) s"Header was not added to ${failed.mkString(",")}. " else ""}${if (oldDateUpdated) "Header date was updated in app/OdController.scala" else ""}"
```
Comment (colin-lamed) [inline src/sbt-test/sbt-auto-build/test-source-generation-older-sbt/build.sbt:20]: OldController spelling

File: src/main/scala/uk/gov/hmrc/SbtAutoBuildPlugin.scala [hunk 3]
```diff
@@ -124,10 +123,29 @@ object PublishSettings {
 // private repo -> only copyright headers
 object HeaderSettings {
 
+  private def retainYearCommentCreator(commentStyle: CommentStyle) = {
+    commentStyle.copy(commentCreator = new CommentCreator() {
+      private val datePattern = "(?s).*?(\\d{4}).*".r
+
+      def findYear(header: String): Option[String] = header match {
```
Comment (colin-lamed) [inline src/main/scala/uk/gov/hmrc/SbtAutoBuildPlugin.scala:130]: mark as `private`?

**Review by colin-lamed — APPROVED** [review]
Excellent!
An entry in the README would be useful

File: src/sbt-test/sbt-auto-build/test-source-generation-older-sbt/build.sbt (hunk 1 above)
Comment (RikSherman) [reply src/sbt-test/sbt-auto-build/test-source-generation-older-sbt/build.sbt:19]: I thought of that but then if both errors occur then only failed.nonEmpty will be reported as sys.error throws an exception.

**Pull request merged**